iroh-net/src/hp/portmapper/upnp.rs

```diff
@@ -0,0 +1,140 @@
+const PORT_MAPPING_LEASE_DURATION_SECONDS: u32 = 0;
+
+/// Maximum duration a UPnP search can take before timing out.
+const SEARCH_TIMEOUT: Duration = Duration::from_secs(1);
+
+/// Tailscale uses the recommended port mapping lifetime for PMP, which is 2 hours. So we assume a
+/// half lifetime of 1h. See <https://datatracker.ietf.org/doc/html/rfc6886#section-3.3>
+const HALF_LIFETIME: Duration = Duration::from_secs(60 * 60);
+
+/// Name with which we register the mapping in the router.
+const PORT_MAPPING_DESCRIPTION: &str = "iroh-portmap";
+
+#[derive(derive_more::Debug, Clone)]
+pub struct Mapping {
+    /// The internet Gateway device (router) used to create this mapping.
+    #[debug("{}", gateway)]
+    gateway: aigd::Gateway,
+    /// The external address obtained by this mapping.
+    external_ip: Ipv4Addr,
+    /// External port obtained by this mapping.
+    external_port: NonZeroU16,
+}
+
+impl Mapping {
+    pub(crate) async fn new(
+        local_addr: Ipv4Addr,
+        port: NonZeroU16,
+        gateway: Option<aigd::Gateway>,
+        preferred_port: Option<NonZeroU16>,
+    ) -> Result<Self> {
+        let local_addr = SocketAddrV4::new(local_addr, port.into());
+
+        // search for a gateway if there is not one already
+        let gateway = if let Some(known_gateway) = gateway {
+            known_gateway
+        } else {
+            aigd::search_gateway(igd::SearchOptions {
+                timeout: Some(SEARCH_TIMEOUT),
+                ..Default::default()
+            })
+            .await?
+        };
+
+        let external_ip = gateway.get_external_ip().await?;
+
+        // if we are trying to get a specific external port, try this first. If this fails, default
+        // to try to get any port
+        if let Some(external_port) = preferred_port {
+            if gateway
+                .add_port(
+                    igd::PortMappingProtocol::UDP,
+                    external_port.into(),
+                    local_addr,
+                    PORT_MAPPING_LEASE_DURATION_SECONDS,
+                    PORT_MAPPING_DESCRIPTION,
+                )
+                .await
+                .is_ok()
+            {
+                return Ok(Mapping {
+                    gateway,
+                    external_ip,
+                    external_port,
+                });
+            }
+        }
+
+        let external_port = gateway
+            .add_any_port(
+                igd::PortMappingProtocol::UDP,
+                local_addr,
+                PORT_MAPPING_LEASE_DURATION_SECONDS,
+                PORT_MAPPING_DESCRIPTION,
+            )
+            .await?
+            .try_into()
+            .map_err(|_| anyhow::anyhow!("upnp mapping got zero external port"))?;
+
+        Ok(Mapping {
+            gateway,
+            external_ip,
+            external_port,
+        })
+    }
+
+    pub fn half_lifetime(&self) -> Duration {
```

Comment:
could be `const`

Reply:
pmp mappings seem to not have predefined lifetimes, so I left it as a function to integrate better with that later on. But at least for upnp yeah it could be used as a constant directly or make the fn const. Which way would you suggest with all that in mind?

Reply:
I guess lets leave it as it is then